Kernel/FileSystem/MountFile.cpp

```diff
@@ -0,0 +1,112 @@
+#include <Kernel/FileSystem/VirtualFileSystem.h>
+#include <Kernel/Memory/PrivateInodeVMObject.h>
+#include <Kernel/Memory/SharedInodeVMObject.h>
+#include <Kernel/Process.h>
+#include <Kernel/StdLib.h>
+
+namespace Kernel {
+
+ErrorOr<NonnullLockRefPtr<MountFile>> MountFile::create(FileSystemInitializer const& file_system_initializer, int flags)
+{
+    // NOTE: We should not open a MountFile if someone wants to either remount or bindmount.
+    // There's a check for this in the fsopen syscall entry handler, but here we just assert
+    // to ensure this never happens.
+    VERIFY(!(flags & MS_BIND));
+    VERIFY(!(flags & MS_REMOUNT));
+    auto mount_specific_data_buffer = TRY(KBuffer::try_create_with_size("Mount Specific Data"sv, 4096, Memory::Region::Access::ReadWrite, AllocationStrategy::AllocateNow));
+    return TRY(adopt_nonnull_lock_ref_or_enomem(new (nothrow) MountFile(file_system_initializer, flags, move(mount_specific_data_buffer))));
+}
+
+MountFile::MountFile(FileSystemInitializer const& file_system_initializer, int flags, NonnullOwnPtr<KBuffer> mount_specific_data)
+    : m_flags(flags)
+    , m_file_system_initializer(file_system_initializer)
+{
+    m_mount_specific_data.with_exclusive([&](auto& our_mount_specific_data) {
+        our_mount_specific_data = move(mount_specific_data);
+        memset(our_mount_specific_data->data(), 0, our_mount_specific_data->size());
+    });
+}
+
+MountFile::~MountFile() = default;
+
+ErrorOr<void> MountFile::ioctl(OpenFileDescription&, unsigned request, Userspace<void*> arg)
+{
+    return m_mount_specific_data.with_exclusive([&](auto& our_mount_specific_data) -> ErrorOr<void> {
+        switch (request) {
+        case MOUNT_IOCTL_SET_MOUNT_SPECIFIC_FLAG: {
+            auto user_mount_specific_data = static_ptr_cast<mount_specific_flag const*>(arg);
+            auto mount_specific_data = TRY(copy_typed_from_user(user_mount_specific_data));
+            if ((mount_specific_data.value_type == mount_specific_flag::ValueType::SignedInteger || mount_specific_data.value_type == mount_specific_flag::ValueType::UnsignedInteger) && mount_specific_data.value_length != 8)
+                return EDOM;
+            if (mount_specific_data.key_string_length > 64)
+                return ENAMETOOLONG;
+            if (mount_specific_data.value_type != mount_specific_flag::ValueType::None && mount_specific_data.value_length == 0)
+                return EINVAL;
+            if (mount_specific_data.value_type != mount_specific_flag::ValueType::None && mount_specific_data.value_addr == nullptr)
+                return EFAULT;
+            if (mount_specific_data.value_type != mount_specific_flag::ValueType::ASCIIString && mount_specific_data.value_length > 64)
+                return E2BIG;
+            if (mount_specific_data.value_type == mount_specific_flag::ValueType::ASCIIString && mount_specific_data.value_length > 1024)
+                return E2BIG;
```

Comment:
Any specific reason for these limitations other than "do not OOM the kernel"? If so, we already copied the data from user memory to kernel memory, so I'm not sure whether this helps much.

Reply:
This is not correct. The actual copying of the string happens in line 92 of that file, for an `ASCIIString` type. We check that the `ASCIIString` is not very long, but the `mount_specific_data` struct is always going to be 28 bytes, exactly.

Reply:
Still, I'm not sure how these limits help anyone. We've had to struggle with plenty of dumb file formats and APIs just because someone thought that `<insert number here>` characters ought to be enough.

Reply:
I agree with @kleinesfilmroellchen - if we have some limits, and they are necessary, then we should either explain them or make them easily configurable by using constants somewhere.

Reply:
Or we could have both - explaining with a comment and having configurable constants :)